Allow up-to 32 servos

IamPete1 opened the pull request:
This allows up to 32 servo outs, it also adds logging for servos 15-32. Tested in SITL by assigning functions and looking in at the new log.

No doubt I have missed some bitmasks somewhere, unfortunately since we have no hardware that supports more than 16 direct outs atm its hard to test the HAL changes. 

Main use case would be for CAN/Sbus servos (although we need to add a second sbus out range for this) so you can free up the physical servo outs.

IamPete1 commented:
I wonder if a we could just assign the PWMs of a board to be the new range in the hwdef to allow to test. Just skipping outputs 1-16.

tridge commented on libraries/SRV_Channel/SRV_Channels.cpp:
```diff
@@ -195,6 +195,102 @@ const AP_Param::GroupInfo SRV_Channels::var_info[] = {
     AP_SUBGROUPINFO(robotis, "_ROB_",  22, SRV_Channels, AP_RobotisServo),
 #endif // HAL_BUILD_AP_PERIPH
 
+#if (NUM_SERVO_CHANNELS >= 17)
```
I think we should have an enable param to get more than 16
Maybe SERVO32_ENABLE ?

magicrub replied:
Possibly.. because this creates a ton of params

IamPete1 replied:
Added `SERVO32_ENABLE` but not enabled it anywhere yet

tridge reviewed (changes requested):
thanks for this!

auturgy commented:
What needs doing in mavlink for this?

> On 12 Jan 2021, at 1:16 pm, Andrew Tridgell <notifications@github.com> wrote:
> 
> ﻿
> @tridge requested changes on this pull request.
> 
> thanks for this!
> 
> In libraries/SRV_Channel/SRV_Channels.cpp:
> 
> > @@ -195,6 +195,102 @@ const AP_Param::GroupInfo SRV_Channels::var_info[] = {
>      AP_SUBGROUPINFO(robotis, "_ROB_",  22, SRV_Channels, AP_RobotisServo),
>  #endif // HAL_BUILD_AP_PERIPH
>  
> +#if (NUM_SERVO_CHANNELS >= 17)
> I think we should have an enable param to get more than 16
> Maybe SERVO32_ENABLE ?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

magicrub commented on libraries/AP_Logger/LogFile.cpp:
```diff
@@ -244,6 +244,41 @@ void AP_Logger::Write_RCOUT(void)
+
+#if NUM_SERVO_CHANNELS >= 15
+    const struct log_RCOUT2 pkt2{
+        LOG_PACKET_HEADER_INIT(LOG_RCOUT2_MSG),
```
This is incorrectly using the same struct name. Perhaps we should call it log_RCOUT15_18?

magicrub commented on libraries/AP_Logger/LogFile.cpp:
```diff
@@ -244,6 +244,41 @@ void AP_Logger::Write_RCOUT(void)
+
+#if NUM_SERVO_CHANNELS >= 19
+    const struct log_RCOUT pkt3{
+        LOG_PACKET_HEADER_INIT(LOG_RCOUT2_MSG),
```
```suggestion
        LOG_PACKET_HEADER_INIT(LOG_RCOUT19_32_MSG),
```

I'd also rather see these all labeled accurately chan19, chan20, chan21... instead of reusing index chan1,chan2,chan3.. It will just lead to confusion

magicrub replied:
Note, I believe the author's intent was to use LOG_RCOUT3_MSG

IamPete1 replied:
Yep, I meant to use `LOG_RCOUT3_MSG`. I figured it was not worth having the duplicate log structure just to have the channel names match.

magicrub commented on libraries/SITL/SITL_Input.h:
```diff
@@ -7,7 +7,7 @@
 */
 struct sitl_input {
-    uint16_t servos[16];
+    uint16_t servos[32];
```
How about we pull from "static const uint8_t max_channels"

IamPete1 replied:
that is in ChibiOS and this is SITL

magicrub commented on libraries/AP_HAL_ChibiOS/RCOutput.h:
```diff
@@ -371,7 +371,7 @@ class ChibiOS::RCOutput : public AP_HAL::RCOutput {
     uint8_t active_fmu_channels;
 
-    static const uint8_t max_channels = 16;
+    static const uint8_t max_channels = 32;
```
```suggestion
#if HAL_MINIMIZE_FEATURES
            static const uint8_t max_channels = 16;
#else
            static const uint8_t max_channels = 32;
#endif
```

magicrub commented:
The idea of this is good for the use case of CAN. However, it doesn't accomplish it for its use case because right now we can't configure servo outputs that aren't Physical. We would need to add that ability to make this worthwhile. @tridge 's suggestion is close, but not quite. Somehow we need to add virtual servo outputs that are independent of hwdef and I don't think it should only be able to add a chunk of 16 more due to the massive param increase.

Perhaps a param that adds virtuals on top of the physical? Of so, each SERVOx param above the physical would need an enable type flag, or at least hidden somehow.

Or am I just talking crazy here and we no longer care about param count now that we have MavFTP?

Also, 32 servos makes for a ton of params on AP_Periph. Perhaps we disable for that build target and only allow physical outputs.

IamPete1 commented:
We could have a enable flag for the second bank of params I guess. But then we would have to make it actually disable the outputs rather than just not showing the params.

IamPete1 commented:
@auturgy this now sends the raw servo outs for 17-32 back over MAVLink using the existing servo raw message, we just use port 1 rather than 0. I would guess all the existing servo override messages will just work, although I have not checked extensively.

IamPete1 commented:
I have forgotten the two BLHeli bitmasks

tridge commented:
@IamPete1 what is the status on this? I think we should aim for this in 4.2.x, which is beta in january

IamPete1 commented:
Rebased and added the BLHeli bitmasks. 

I think the main stumbling block for this is that it needs full 32 bit params, maybe this works over FTP? But I don't think does via the old param transport. 

The other option would be to split all the mask params into two 16 bits. Maybe some special handling like we do for Vector3 params? We will have to do something like that if we ever want to goto more than 32 outputs in anycase....

IamPete1 commented:
Size compare is interesting, I seem to have moved stuff from bss to text, I have no idea if that matters or not...

```
Binary           text             data         bss               total
---------------  ---------------  -----------  ----------------  -----------
antennatracker   872 (+0.1328%)   0 (0.0000%)  -872 (-0.1912%)   0 (0.0000%)
arducopter-heli  876 (+0.0951%)   0 (0.0000%)  -876 (-0.4641%)   0 (0.0000%)
ardurover        1032 (+0.1237%)  0 (0.0000%)  -1032 (-0.3704%)  0 (0.0000%)
arduplane        1116 (+0.1147%)  0 (0.0000%)  -1116 (-0.8153%)  0 (0.0000%)
blimp            916 (+0.1357%)   0 (0.0000%)  -916 (-0.2094%)   0 (0.0000%)
arducopter       864 (+0.0923%)   0 (0.0000%)  -864 (-0.4965%)   0 (0.0000%)
ardusub          868 (+0.1078%)   0 (0.0000%)  -868 (-0.2825%)   0 (0.0000%)
```

andyp1per commented:
bss is fake space as its all zeros so doesn't take up space until it is loaded in memory, so text is the only thing that matters

lthall commented:
Can we have some sort of variable that would allow us to use 17 to 32 on the sbus output?

IamPete1 commented on libraries/AP_Logger/LogFile.cpp:
```diff
@@ -197,6 +197,41 @@ void AP_Logger::Write_RCOUT(void)
     };
     WriteBlock(&pkt, sizeof(pkt));
+
+#if NUM_SERVO_CHANNELS >= 15
```
done

tridge commented:
rebased and added more missing 32 bit handling

tridge commented:
added SERVO_32_ENABLE parameter

tridge commented:
@IamPete1 I added a SERVO_32_ENABLE and made the output loop not look at servos above 16 if disabled

peterbarker commented on libraries/AP_Logger/LogFile.cpp:
```diff
@@ -178,25 +178,68 @@ void AP_Logger::Write_RCIN(void)
+
+#if NUM_SERVO_CHANNELS >= 15
+    if ((enabled_mask & 0x3C000) != 0) {
+        const struct log_RCOUT2 pkt2{
+            LOG_PACKET_HEADER_INIT(LOG_RCOUT2_MSG),
```
Not a blocker, but we've tended to use the message name in these.  So `log_RCO2` and `LOG_RCO2_MSG`

peterbarker commented on libraries/SRV_Channel/SRV_Channel_aux.cpp:
```diff
@@ -587,15 +595,15 @@ float SRV_Channels::get_slew_limited_output_scaled(SRV_Channel::Aux_servo_functi
         return functions[function].channel_mask;
     }
-    return 0;
+    return invalid_mask;
```
What's this patch for?

tridge replied:
it is needed for this line:
const uint32_t enabled_mask = ~SRV_Channels::get_output_channel_mask(SRV_Channel::k_GPIO);
in AP_Logger::Write_RCOUT()

peterbarker commented on libraries/SRV_Channel/SRV_Channels.cpp:
```diff
@@ -236,12 +237,117 @@ const AP_Param::GroupInfo SRV_Channels::var_info[] = {
     // @DisplayName: Servo GPIO mask
-    // @Description: This sets a bitmask of outputs which will be available as GPIOs. Any auxiliary output with either the function set to -1 or with the corresponding bit set in this mask will be available for use as a GPIO pin
-    // @Bitmask: 0:Servo 1, 1:Servo 2, 2:Servo 3, 3:Servo 4, 4:Servo 5, 5:Servo 6, 6:Servo 7, 7:Servo 8, 8:Servo 9, 9:Servo 10, 10:Servo 11, 11:Servo 12, 12:Servo 13, 13:Servo 14, 14:Servo 15, 15:Servo 16
+    // @Description: This sets a bitmask of outputs which will be available as GPIOs. Any auxillary output with either the function set to -1 or with the corresponding bit set in this mask will be available for use as a GPIO pin
```
Reverted the spelling fix PR :-)

tridge replied:
fixed

peterbarker commented on libraries/SRV_Channel/SRV_Channels.cpp:
```diff
@@ -236,12 +237,117 @@ const AP_Param::GroupInfo SRV_Channels::var_info[] = {
-    // @Description: This sets a bitmask of outputs which will be available as GPIOs. Any auxiliary output with either the function set to -1 or with the corresponding bit set in this mask will be available for use as a GPIO pin
-    // @Bitmask: 0:Servo 1, 1:Servo 2, 2:Servo 3, 3:Servo 4, 4:Servo 5, 5:Servo 6, 6:Servo 7, 7:Servo 8, 8:Servo 9, 9:Servo 10, 10:Servo 11, 11:Servo 12, 12:Servo 13, 13:Servo 14, 14:Servo 15, 15:Servo 16
+    // @Description: This sets a bitmask of outputs which will be available as GPIOs. Any auxillary output with either the function set to -1 or with the corresponding bit set in this mask will be available for use as a GPIO pin
+    // @Bitmask: 0:Servo 1, 1:Servo 2, 2:Servo 3, 3:Servo 4, 4:Servo 5, 5:Servo 6, 6:Servo 7, 7:Servo 8, 8:Servo 9, 9:Servo 10, 10:Servo 11, 11:Servo 12, 12:Servo 13, 13:Servo 14, 14:Servo 15, 15:Servo 16, 16:Servo 17, 17:Servo 18, 18:Servo 19, 19:Servo 20, 20:Servo 21, 21:Servo 22, 22:Servo 23, 23:Servo 24, 24:Servo 25, 25:Servo 26, 26:Servo 27, 27:Servo 28, 28:Servo 29, 29:Servo 30, 30:Servo 31 ...
```
Should we actually be nominating the high bits here?  We can't transport these with the traditional mavlink protocol.

peterbarker reviewed (commented):
A name for the mask type would be good (even if only by typedef initially)....

I didn't spot anything wrong, but don't have time to test here ATM.

geofrancis commented:
PCA9685 support in the regular firmware would make plugging in more servos much simpler even if they are limited to basic servo output and low update speeds, it would be good for non-critical devices. its already supported under linux hal.
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_Linux/RCOutput_PCA9685.h

tridge commented:
@geofrancis indeed, that would be nice, although we should not hold this PR for that
main use case for now is CAN

tridge commented:
I have tested on a CubeOrange with this in hwdef.dat:
```
PE14 TIM1_CH4 TIM1 PWM(9) GPIO(50)
PE13 TIM1_CH3 TIM1 PWM(10) GPIO(51)
PE11 TIM1_CH2 TIM1 PWM(11) GPIO(52)
PE9  TIM1_CH1 TIM1 PWM(12) GPIO(53)
PD13 TIM4_CH2 TIM4 PWM(13) GPIO(54)
PD14 TIM4_CH3 TIM4 PWM(14) GPIO(55)
```
servos 17 to 22 work fine on AUX pins
I also tested with a MatekL431-DShot adapter and got servos 17 to 19 working over CAN

tridge commented:
had a successful flight on 2 planes. One CubeOrange with hwdef.dat changed so aux servos are 17-22. Flew with VTOL motors on 17 to 20 and fwd motor on 21. All good!

IamPete1 commented:
@geofrancis If your on a 2M board you should see `SERVO_32_ENABLE`
